topo/node/ceos/ceos.go

```diff
@@ -66,99 +76,143 @@ func New(nodeImpl *node.Impl) (node.Node, error) {
-
-	return err
+	if err := n.CreateCRD(ctx); err != nil {
+		return fmt.Errorf("node %s failed to create custom resource definition %w", n.Name(), err)
+	}
+	return nil
 }
 
-func (n *Node) GenerateSelfSigned(ctx context.Context) error {
-	selfSigned := n.Proto.GetConfig().GetCert().GetSelfSigned()
-	if selfSigned == nil {
-		log.Infof("%s - no cert config", n.Name())
-		return nil
+func (n *Node) CreateCRD(ctx context.Context) error {
+	log.Infof("Creating new CEosLabDevice CRD for node: %v", n.Name())
+	proto := n.GetProto()
+	config := proto.GetConfig()
+	device := &ceos.CEosLabDevice{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "ceoslab.arista.com/v1alpha1",
+			Kind:       "CEosLabDevice",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      n.Name(),
+			Namespace: n.GetNamespace(),
+			Labels: map[string]string{
+				"app":  n.Name(),
+				"topo": n.GetNamespace(),
+			},
+		},
+		Spec: ceos.CEosLabDeviceSpec{
+			EnvVar:        config.GetEnv(),
+			Image:         config.GetImage(),
+			Args:          config.GetArgs(),
+			Resources:     proto.GetConstraints(),
+			NumInterfaces: uint32(len(proto.GetInterfaces())),
+			Sleep:         config.GetSleep(),
+		},
 	}
-	log.Infof("%s - generating self signed certs", n.Name())
-	log.Infof("%s - waiting for pod to be running", n.Name())
+	for label, v := range proto.GetLabels() {
+		device.ObjectMeta.Labels[label] = v
+	}
+	for port, service := range proto.GetServices() {
+		if device.Spec.Services == nil {
+			device.Spec.Services = map[string]ceos.ServiceConfig{}
+		}
+		device.Spec.Services[service.Name] = ceos.ServiceConfig{
+			TCPPorts: []uint32{port},
+		}
+	}
+	if cert := config.GetCert(); cert != nil {
+		if ssCert := cert.GetSelfSigned(); ssCert != nil {
+			certConfig := ceos.CertConfig{
+				SelfSignedCerts: []ceos.SelfSignedCertConfig{{
+					CertName:   ssCert.CertName,
+					KeyName:    ssCert.KeyName,
+					KeySize:    ssCert.KeySize,
+					CommonName: ssCert.CommonName,
+				}},
+			}
+			device.Spec.CertConfig = certConfig
+		}
+	}
+	for k, v := range n.GetProto().GetInterfaces() {
+		if device.Spec.IntfMapping == nil {
+			device.Spec.IntfMapping = map[string]string{}
+		}
+		device.Spec.IntfMapping[k] = v.GetName()
+	}
+	// Post to k8s
+	r := ceos.CEosLabDevice{}
+	err := n.KubeClient.CoreV1().RESTClient().
+		Post().
+		AbsPath("/apis/" + API).
+		Namespace(n.Namespace).
+		Resource(RESOURCE).
+		Body(device).
+		Do(ctx).
+		Into(&r)
+	// Wait for pods
 	w, err := n.KubeClient.CoreV1().Pods(n.Namespace).Watch(ctx, metav1.ListOptions{
-		FieldSelector: fields.SelectorFromSet(
-			fields.Set{metav1.ObjectNameField: n.Name()},
-		).String(),
+		FieldSelector: fields.SelectorFromSet(fields.Set{metav1.ObjectNameField: n.Name()}).String(),
 	})
 	if err != nil {
 		return err
 	}
 	for e := range w.ResultChan() {
 		p := e.Object.(*corev1.Pod)
-		if p.Status.Phase == corev1.PodRunning {
+		if p.Status.Phase == corev1.PodPending {
```

Comment:
Why is the behavior changed to wait for Pending instead of Running?

Reply:
Thanks for catching this. I install a startup probe in the operator and I was experimenting with blocking on that, but it doesn't appear the core API exports that state as directly (would have had to iterate over pod conditions) and I decided not to move forward with it. It would also have had the side effect of not allowing pods to "boot" concurrently. I must have botched it when going back.

Reply:
As we discussed not sure how I got so confused about this, but this was originally PodPending. Setting it to PodRunning introduced a deadlock whereby init containers would be waiting for pods whose creation was blocked here.